make --run-specified-test-case use --include

walterddr opened the pull request:
instead of having specific logic to handle run-specific-test-case, we provide the flag to override include or bring-to-front with the SPECIFIED_TEST_CASES_FILE.

janeyx99 commented on test/run_test.py:
```diff
@@ -401,8 +401,8 @@
-# options.run_specified_test_cases is enabled.
-SPECIFIED_TEST_CASES_FILE: str = '.pytorch_specified_test_cases.csv'
+# options.use_specified_test_cases is enabled.
+SPECIFIED_TEST_CASES_FILE: str = '.pytorch_use_specified_test_cases.csv'
```
Why do we want to change the name of the csv?

walterddr replied:
oops........ regex search/replace error. undoing

janeyx99 commented on test/run_test.py:
```diff
@@ -834,11 +834,11 @@ def parse_args():
-        help='runs specified test cases from previous OSS CI stats from a file, format CSV',
+        choices=['include', 'bring-to-front'],
+        help='if this flag is set. run_test will load specified test cases file from '
+        'SPECIFIED_TEST_CASES_FILE and use the list to determine what test should be run',
```
So we'd just force users to always name their file to SPECIFIED_TEST_CASES_FILE? Is there a way to do this where we still allow the user to specify their own custom file name?

walterddr replied:
i don't think `run_test.py` is for user directly - since developer should directly use `test/test_*.py` which calls common_utils.run_test() or other customized handlers. Do you think we have use case when user will do this?

janeyx99 commented on test/run_test.py:
```diff
@@ -834,11 +834,11 @@ def parse_args():
         type=str,
-        const=SPECIFIED_TEST_CASES_FILE,
-        help='runs specified test cases from previous OSS CI stats from a file, format CSV',
+        choices=['include', 'bring-to-front'],
```
I think the 'include' is a little misleading, I think it suggests we run everything and we make sure to include this subset of tests, whereas the actual functionality would be to _only_ run these tests, right? Maybe something like include-only or like, exclude-others or something. 

I also think we should document what these choices mean in the help.

walterddr replied:
i see your concern. `-i` and `--include` option is clearly indicating it is only running those. but not at this option. I can improve the docstring

janeyx99 commented on test/run_test.py:
```diff
@@ -1102,7 +1103,7 @@ def export_S3_test_times(test_times_filename: str, test_times: Dict[str, float])
 
 
-def load_specified_test_cases(filename: str) -> None:
+def load_use_specified_test_cases(filename: str) -> None:
```
I think the name change here is not better than the original, was this an unintentional find/replace?

walterddr replied:
yup. unintentional find/replace

janeyx99 commented on test/run_test.py:
```diff
@@ -893,6 +893,11 @@ def exclude_tests(exclude_list, selected_tests, exclude_message=None):
+    if options.use_specified_test_cases == 'include':
+        options.include = list(SPECIFIED_TEST_CASES_DICT.keys())
+    elif options.use_specified_test_cases == 'bring-to-front':
+        options.bring_to_front = list(SPECIFIED_TEST_CASES_DICT.keys())
```
So in get_test_case_args, I print a warning in the case where a test_module is passed in but it isn't in SPECIFIED_TEST_CASES_DICT. Could you either rephrase that comment since now, other test modules would still be allowed?

I'm also hesitant about this option currently, because say we have "test_nn": ["test1", "test1324"] in the dictionary. This option might mislead people into thinking that we will run the other tests in test_nn after we run test1 and test1324, but that's not true. The other test cases would not be run at all.

janeyx99 reviewed (commented):
This is definitely going in the right direction, though I'm not sure this change would make the most sense just yet (particularly with the bring_to_front option).

Commit:
let run-specified-test-cases option use existing logic

allow include or bring-to-front option
marked future to do
address diff comments

walterddr commented on test/run_test.py:
```diff
@@ -394,10 +394,10 @@
 # For example:
 # {
-#   "test_nn": ["test_doubletensor_avg_pool3d", "test_share_memory", "test_hook_requires_grad"],
+#   "`test_nn`": ["test_doubletensor_avg_pool3d", "test_share_memory", "test_hook_requires_grad"],
```
```suggestion
#   "test_nn": ["test_doubletensor_avg_pool3d", "test_share_memory", "test_hook_requires_grad"],
```

walterddr committed:
Update test/run_test.py

walterddr committed:
Update test/run_test.py

Commit:
support __all__ and improve help message

janeyx99 commented on test/run_test.py:
```diff
@@ -532,10 +532,13 @@ def get_slow_tests_based_on_S3() -> List[str]:
-        sys.exit(f'Warning! Test module {test_module} is not found in the specified tests dict. This should never'
-                 'happen as we make a check for that before entering this function.')
     args = []
+    if test_module not in SPECIFIED_TEST_CASES_DICT:
```
maybe clarify by adding a comment that this should be triggered in the bring_to_front case/if you don't want to make the distinction, I think this if statement could be merged with the second

janeyx99 commented on test/run_test.py:
```diff
@@ -838,8 +843,23 @@ def parse_args():
+        ' ...\n'
+        ' <test_module>,__all__\n'
+        ' ...\n'
+        'file content is used based on option "--use-specified-test-cases-for".'
```
```suggestion
        'how we use the stats will be based on option "--use-specified-test-cases-by".'
```
I think this could help clarify

janeyx99 commented on test/run_test.py:
```diff
@@ -838,8 +843,23 @@ def parse_args():
+        'file content is used based on option "--use-specified-test-cases-for".'
+    )
+    parser.add_argument(
+        '--use-specified-test-cases-for',
```
```suggestion
        '--use-specified-test-cases-by',
```

janeyx99 commented on test/run_test.py:
```diff
@@ -838,8 +843,23 @@ def parse_args():
+        choices=['include', 'bring-to-front'],
+        default='include',
+        help='used together with option "--run-specified-test-cases". When specified test case '
+        'file is set. This option allow user to control whether to only run the specified test '
+        'modules or to also run the remaining modules. Note: regardless of this option, we '
+        'only run the specified test cases within a test module.',
```
```suggestion
        'file is set, this option allows the user to control whether to only run the specified test '
        'modules or to simply bring the specified modules to front and also run the remaining '
        'modules. Note: regardless of this option, we will only run the specified test cases '
        ' within a specified test module. For unspecified test modules with the bring-to-front '
        'option, all test cases will be run, as one may expect.',
```

janeyx99 reviewed (approved):
I think this is good now! Left clarification nits.

I think we should change `--use-specified-test-cases-for` to `--use-specified-test-cases-by` to make the options more different/clear as to what they do.

walterddr commented on test/run_test.py:
```diff
@@ -532,10 +532,13 @@ def get_slow_tests_based_on_S3() -> List[str]:
-        sys.exit(f'Warning! Test module {test_module} is not found in the specified tests dict. This should never'
-                 'happen as we make a check for that before entering this function.')
     args = []
+    if test_module not in SPECIFIED_TEST_CASES_DICT:
+        return args
+
+    # if specified with __all__ then run all tests
+    if '__all__' in SPECIFIED_TEST_CASES_DICT[test_module]:
+        return args
```
```suggestion
    # if test_module not specified or specified with '__all__' then run all tests
    if test_module not in SPECIFIED_TEST_CASES_DICT or '__all__' in SPECIFIED_TEST_CASES_DICT[test_module]:
        return args
```

walterddr committed:
Apply suggestions from code review

Co-authored-by: Jane (Yuan) Xu <31798555+janeyx99@users.noreply.github.com>
